File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow, fast = nums[0], nums[0]
        n = len(nums)

        while fast < n and nums[fast] < n:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                slow = nums[0]
                while slow != fast:
                    slow = nums[slow]
                    fast = nums[fast]
                return slow

        return -1

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int slow = nums[0], fast = nums[0];
        int n = nums.length;

        while (fast < n && nums[fast] < n) {
            slow = nums[slow];
            fast = nums[nums[fast]];
            if (slow == fast) {
                slow = nums[0];
                while (slow != fast) {
                    slow = nums[slow];
                    fast = nums[fast];
                }
                return slow;
            }
        }

        return -1;
    }
}